### PR TITLE
Upgrade to terraform 0.12

### LINKED ISSUE #2: Remove hardcoded memcached in security group name
Security group name is currently set to `"${var.prefix}-memcached-sg` but the module supports both memcached and redis hence we should remove it from the name.

### PR DESCRIPTION
**Pull request by mikael-lindstrom**
Small breaking change that fixes #2, other then that it should be backwards compatible.

### CODE REVIEW

File: outputs.tf [hunk 1]
```diff
@@ -2,18 +2,22 @@
...
 output "cluster_address" {
-  value = "${aws_elasticache_cluster.main.cluster_address}"
+  description = "The DNS name of the cache cluster without the port appended. Only available for memcached"
+  value       = var.engine == "memcached" ? aws_elasticache_cluster.main.cluster_address : ""
 }
 
 output "security_group_id" {
   description = "The ID of the security group."
-  value       = "${element(concat(aws_security_group.main.*.id, list("")), 0)}"
+  value       = element(concat(aws_security_group.main.*.id, [""]), 0)
```
Comment (itsdalmo) [inline outputs.tf]: Don't need the `concat` here anymore, just use `[0]`.

Comment (itsdalmo) [reply outputs.tf]: Oh, and `[*]` for the splat syntax. Please verify that I'm correct in that `[*]` is the future:
https://www.terraform.io/docs/configuration/expressions.html#splat-expressions

File: variables.tf [hunk 2]
```diff
@@ -12,7 +12,7 @@ variable "subnet_group_name" {
 }
 
 variable "parameter_group_name" {
-  default = "default.memcached1.4"
+  default = "default.memcached1.5"
```
Comment (itsdalmo) [inline variables.tf]: This would be a breaking change (desired state changes and requires replacement) no?

File: variables.tf [hunk 3]
```diff
@@ -41,11 +41,12 @@ variable "vpc_id" {
 
 variable "subnet_ids" {
   description = "ID of subnets for the RDS subnet group."
-  type        = "list"
+  type        = list(string)
```
Comment (itsdalmo) [inline variables.tf:59]: I would also suggest adding type info to _all_ variables when we are at it. So `type = string`, `bool`, `number` etc.

File: outputs.tf [hunk 4]
```diff
@@ -2,18 +2,22 @@
 # https://www.terraform.io/docs/providers/aws/r/elasticache_cluster.html
 
 output "cache_nodes" {
-  value = "${aws_elasticache_cluster.main.cache_nodes}"
+  description = "List of node objects including id, address, port and availability_zone."
```
Comment (itsdalmo) [inline outputs.tf:5]: Loving it!

**Review by itsdalmo — COMMENTED** [review]
Looking good! Just a couple of nitpicks!

File: variables.tf (hunk 2 above)
Comment (mikael-lindstrom) [reply variables.tf]: This is due to the default version for memcached changing. We could lock `engine_version` to 1.4 but that might change desired state too.

Comment (mikael-lindstrom) [reply variables.tf]: Should we leave this in and document the changes?

Comment (itsdalmo) [reply variables.tf]: I think that sounds reasonable, if default version has changed then it will be a breaking change either way right?

Comment (mikael-lindstrom) [reply variables.tf]: Yepp it will, I'm going to leave it in and make note in the release.

**Review by itsdalmo — APPROVED** [review]
LGTM 💯

**Pull request merged**